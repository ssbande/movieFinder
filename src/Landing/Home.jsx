import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Container } from 'semantic-ui-react';
import { connect } from 'react-redux';

import InitRender from './Init';
import ListData from './ListData';
import ListMenu from './ListMenu';
import {
  getFilter,
  getInitialMenu,
  getMenus,
  getMovies,
  getShows,
  getArtists,
  shouldShowStartScreen
} from '../state/selectors';
import constants from '../utils/constants';
import { FilterProps } from '../utils/propTypes';

import './data.css';

export const Home = props => {
  const { showStart, filter, initMenu, availableMenus, ...restProps } = props;
  const [activeItem, setActiveItem] = useState(initMenu);

  useEffect(() => { setActiveItem(initMenu) }, [initMenu])

  const handleItemClick = (e, { name }) => {
    const menu = Object.values(constants.config).find(item => item.label === name);
    setActiveItem(menu.route);
  }

  return <Container>
    {showStart
      ? <InitRender />
      : (
        <main>
          <ListMenu activeItem={activeItem} handleItemClick={handleItemClick} filter={filter} menus={availableMenus} />
          <ListData activeItem={activeItem} {...restProps} />
        </main>
      )}
  </Container>
}

Home.propTypes = {
  filter: FilterProps,
  movies: PropTypes.array,
  shows: PropTypes.array,
  artists: PropTypes.array,
  showStart: PropTypes.bool.isRequired,
  initMenu: PropTypes.string.isRequired,
  availableMenus: PropTypes.array
}

const mapState = state => ({
  filter: getFilter(state),
  movies: getMovies(state),
  shows: getShows(state),
  artists: getArtists(state),
  showStart: shouldShowStartScreen(state),
  initMenu: getInitialMenu(state),
  availableMenus: getMenus(state),
})

export default connect(mapState)(Home);