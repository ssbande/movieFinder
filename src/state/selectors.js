import get from 'lodash/get';
import constants from '../utils/constants';
const { config } = constants;

export const getQuery = state => get(state, 'userQuery', null);
export const getFilter = state => get(state, 'filter', null);
export const getSubmitted = state => get(state, 'submitted', false);
export const getSuggestions = state => get(state, 'suggestions', []);
export const getMovies = state => get(state, 'movies.results', []);
export const getShows = state => get(state, 'shows.results', []);
export const getArtists = state => get(state, 'artists.results', []);
export const getDetailInfo = state => get(state, 'detailInfo', null);
export const getError = state => get(state, 'error', false);

export const getInitialMenu = state => {
  const filter = getFilter(state);
  return filter === config.all.category
    ? config.movie.route
    : config[filter].route;
}

export const getMenus = state => {
  const filter = getFilter(state);
  return Object.entries(config)
    .filter(item => {
      return (filter !== 'all') 
        ? item[0] === filter
        : item[0] !== 'all';
    });
}

export const shouldShowSuggestions = state => {
  const { queryLengthToStartSuggestions } = constants;
  const query = getQuery(state);
  const isSubmitted = getSubmitted(state);

  return query.length >= queryLengthToStartSuggestions && !isSubmitted;
}

export const shouldShowStartScreen = state => {
  const movies = getMovies(state);
  const shows = getShows(state);
  const artists = getArtists(state);
  const submitted = getSubmitted(state);

  return !movies.length && !shows.length && !artists.length && !submitted;
}