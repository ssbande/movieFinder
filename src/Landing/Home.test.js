import React from 'react';
import { shallow } from 'enzyme';
import { makeData } from '../state/selectors.test';

import { Home } from './Home';

const setup = () => ({
  props: {
    filter: 'all',
    movies: makeData('movie'),
    shows: makeData('show'),
    artists: makeData('artist'),
    showStart: false,
    initMenu: 'movies',
    handleClick: jest.fn(),
    availableMenus: [[
      'movies',
      {
        category: 'movie',
        route: 'movies',
        label: 'Movies',
        order: 1,
        successType: 'MOVIE_SUCCESS'
      }
    ]]
  }
})

describe('Home page tests', () => {
  it('should render correctly', () => {
    const { props } = setup();
    const wrapper = shallow(<Home {...props} />);
    expect(wrapper).toMatchSnapshot();
  })
})