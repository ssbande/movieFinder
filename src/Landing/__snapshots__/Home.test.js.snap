// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page tests should render correctly 1`] = `ShallowWrapper {}`;
